server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RecipientRewriteTableProcessorTest.java

```diff
@@ -0,0 +1,331 @@
+        mail = FakeMail.builder()
+                .mimeMessage(message)
+                .build();
+        processor.replaceReturnPath(mail);
+        
+        assertThat(mail.getMessage().getHeader(RFC2822Headers.RETURN_PATH)).containsOnly("<>");
+    }
+
+    @Test
+    public void replaceReturnPathShouldRewriteMessageHeaderWhenMailSenderNotNull() throws Exception {
+        mail = FakeMail.builder()
+                .mimeMessage(message)
+                .sender(MailAddressFixture.ANY_AT_JAMES)
+                .build();
+        processor.replaceReturnPath(mail);
+        
+        assertThat(mail.getMessage().getHeader(RFC2822Headers.RETURN_PATH)).containsOnly("<any@james.apache.org>");
+    }
+
+    @Test
+    public void logAndAddErrorShouldWriteLogAndAddMailAddressIntoErrorList() {
+        Builder<MailAddress> errors = ImmutableList.builder();
+        Exception e = mock(Exception.class);
+        processor.logAndAddError(errors, MailAddressFixture.ANY_AT_JAMES, e, mockMailetContext);
+        
+        verify(mockMailetContext, times(1)).log(eq(LogLevel.INFO), anyString(), eq(e));
+        assertThat(errors.build()).containsOnly(MailAddressFixture.ANY_AT_JAMES);
+    }
+    
+    @Test
+    public void rewriteRecipientsShouldReturnNewListOfRecipients() throws Exception {
+        mail = FakeMail.builder()
+                .mimeMessage(message)
+                .recipient(MailAddressFixture.ANY_AT_JAMES)
+                .build();
+        
+        ImmutableList<MailAddress> newRecipients = processor.rewriteRecipients(mail, getProcessFunction());
+        
+        assertThat(newRecipients).containsExactly(MailAddressFixture.OTHER_AT_JAMES, MailAddressFixture.ANY_AT_JAMES2);
+    }
+
+    @Test
+    public void rewriteRecipientsShouldReturnNewFullListOfRecipients() throws Exception {
+        mail = FakeMail.builder()
+                .mimeMessage(message)
+                .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.ANY_AT_LOCAL)
+                .build();
+        
+        ImmutableList<MailAddress> newRecipients = processor.rewriteRecipients(mail, getProcessFunction());
+        
+        assertThat(newRecipients).containsExactly(MailAddressFixture.OTHER_AT_JAMES, MailAddressFixture.ANY_AT_JAMES2, MailAddressFixture.ANY_AT_JAMES);
+    }
+
+    @Test
+    public void rewriteRecipientsShouldReturnRightListOfRecipients() throws Exception {
+        mail = FakeMail.builder()
+                .mimeMessage(message)
+                .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.ANY_AT_LOCAL, MailAddressFixture.OTHER_AT_JAMES)
+                .build();
+        
+        ImmutableList<MailAddress> newRecipients = processor.rewriteRecipients(mail, getProcessFunction());
+        
+        assertThat(newRecipients).containsExactly(MailAddressFixture.OTHER_AT_JAMES, MailAddressFixture.ANY_AT_JAMES2, MailAddressFixture.ANY_AT_JAMES);
```

Comment:
To "containsOnly" please. Maybe we don't care about order but we care about extra entries ;-)